### PR TITLE
Environment Configuration task

### PR DESCRIPTION
**Pull request by chelnak**

### CODE REVIEW

File: Build-Task.ps1 [hunk 1]
```diff
@@ -0,0 +1,245 @@
...
+    $null = New-Item -Path $PackageTemp -ItemType Directory -Force
+    Write-Verbose -Message "ResolvedTaskRoot: $ResolvedTaskRoot"
+    Write-Verbose -Message "ReleaseTaskRoot: $ReleaseTaskRoot"
+    Write-Verbose -Message "ConfigPath: $ConfigPath"
+
+    if (!$ConfigPath) {
+        throw "Could not find dependency.json at $ConfigPath"
+    }
+
+    Write-Verbose -Message "Retrieving config deinition from $ConfigPath"
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    Write-Verbose -Message "Retrieving config definition from $ConfigPath"
```

File: tasks/EnvironmentConfiguration/task/task.json [hunk 2]
```diff
@@ -0,0 +1,92 @@
+{
+  "id": "7c21ef57-08a3-4dab-a647-1c790f76ad19",
+  "name": "GenerateEnvironmentConfiguration",
+  "friendlyName": "Generate Configuration",
+  "description": "Generate, validate and inject custom application configuration in to table storage.",
+  "helpMarkDown": "",
+  "category": "Utility",
+  "author": "Craig Gumbley",
```
Comment (EwanNoble) [inline tasks/EnvironmentConfiguration/task/task.json]: This should probably match the publisher account

Comment (chelnak) [reply tasks/EnvironmentConfiguration/task/task.json]: Author? Hasn’t changed since v1 but agree.

**Comment by chelnak** [comment]
@EwanNoble comments addressed.

@jack-education @dalemorson - please can you review?

**Comment by chelnak** [comment]
Agreed with @EwanNoble that this can be merged to master with the caveat of #5.

@jack-education @dalemorson Please can you look over this.

File: Build-Task.ps1 [hunk 3]
```diff
@@ -0,0 +1,244 @@
...
+    [Parameter(Mandatory = $false, ParameterSetName = "Publish")]
+    [switch]$Publish,
+    [Parameter(Mandatory = $true, ParameterSetName = "Publish")]
+    [string]$AccessToken,
+    [Parameter(Mandatory = $false)]
+    [switch]$SkipRestore
+)
+
+# Default error action to stop
+$ErrorActionPreference = "Stop"
```
Comment (dalemorson) [inline Build-Task.ps1]: Are you happy with using a global error action?

File: Build-Task.ps1 [hunk 4]
```diff
@@ -0,0 +1,244 @@
...
+
+        foreach ($Package in $Config.Include | Sort-Object -Property Type) {
+            Write-Verbose -Message "Processing package dependency $($Package.Name)"
+            Write-Verbose -Message "Clean package directories: $($NoResotre.IsPresent)"
+
+            Write-Verbose -Message "Resolving package path"
+            [System.IO.FileInfo]$ResolvedPackagePath = "$($ReleaseTaskRoot)/$($Package.Path)"
+            Write-Verbose -Message "ResolvedPackagePath: $ReleaseTaskRoot"
+
+
```
Comment (dalemorson) [inline Build-Task.ps1]:
```suggestion
```

Comment (dalemorson) [inline Build-Task.ps1]:
```suggestion
```

File: Build-Task.ps1 [hunk 5]
```diff
@@ -0,0 +1,244 @@
...
+                }
+                'Local' {
+                    if ($Package.Copy) {
+                        $Package.Copy | ForEach-Object {
+                            Write-Host "[Local] Copying dependency $_ to $($Package.Path)"
+                            Copy-Item -Path $ResolvedTaskRoot/$_ -Destination $ResolvedPackagePath -Recurse -Force
+                        }
+                    }
+                }
+                'Defaut' {
```
Comment (dalemorson) [inline Build-Task.ps1]:
```suggestion
                'Default' {
```

File: Build-Task.ps1 [hunk 6]
```diff
@@ -0,0 +1,244 @@
+<#
+    .SYNOPSIS
+    Install common task modules at build time.
```
Comment (jack-education) [inline Build-Task.ps1]:
```suggestion
    Install common task modules at build time
```

File: tasks/EnvironmentConfiguration/dependency.json [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+  "include": [
+    {
+      "type": "PSGallery",
+      "name": "VstsTaskSdk",
+      "version": "0.11.0",
+      "copy":[
+        "0.11.0"
+      ],
+      "path": "task/ps_modules"
+
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/dependency.json]:
```suggestion
```

File: tasks/EnvironmentConfiguration/dependency.json [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+      ],
+      "path": "task/lib"
+    },
+    {
+      "type": "GitHub",
+      "name": "Microsoft/azure-pipelines-tasks",
+      "copy": [
+        "Tasks/Common/VstsAzureHelpers_",
+        "Tasks/Common/TlsHelper_"
+
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/dependency.json]:
```suggestion
```

File: Build-Task.ps1 [hunk 9]
```diff
@@ -0,0 +1,244 @@
+<#
+    .SYNOPSIS
+    Install common task modules at build time
+
+    .DESCRIPTION
+    Install common task modules at build time.
```
Comment (jack-education) [inline Build-Task.ps1]:
```suggestion
    Install common task modules at build time
```

File: Build-Task.ps1 [hunk 10]
```diff
@@ -0,0 +1,244 @@
...
+    - git
+    - tfx-cli
+#>
+[CmdletBinding(DefaultParameterSetName = "Build")]
+Param(
+    [Parameter(Mandatory = $true)]
+    [System.Io.FileInfo]$TaskRoot,
+    [Parameter(Mandatory = $false)]
+    [switch]$Clean,
+    [Parameter(Mandatory = $false)]
```
Comment (jack-education) [inline Build-Task.ps1]:
```suggestion
```

File: tasks/EnvironmentConfiguration/task/ps_modules/Handlers/public/New-ConfigurationTableEntry.ps1 [hunk 11]
```diff
@@ -0,0 +1,61 @@
...
+        $SchemaPath = "$($SourcePath)$($TargetFilename)"
+        $Schemas = Get-ChildItem -Path $SchemaPath -File -Recurse
+
+        foreach ($Schema in $Schemas) {
+
+            $Configuration = Build-ConfigurationEntity -SchemaDefinitionPath $Schema.FullName
+            Test-ConfigurationEntity -Configuration $Configuration -SchemaDefinitionPath $Schema.FullName
+
+            $NewEntityParameters = @{
+
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/task/ps_modules/Handlers/public/New-ConfigurationTableEntry.ps1]:
```suggestion
```

File: tasks/EnvironmentConfiguration/task/ps_modules/Handlers/private/storage/Set-TableEntity.ps1 [hunk 12]
```diff
@@ -0,0 +1,29 @@
...
+            $null = $StorageTable.CloudTable.Execute([Microsoft.WindowsAzure.Storage.Table.TableOperation]::InsertOrReplace($Entity))
+        }
+
+    }
+    catch {
+        Write-Error -Message "$_" -ErrorAction Stop
+    }
+    finally {
+        Trace-VstsLeavingInvocation $MyInvocation
+
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/task/ps_modules/Handlers/private/storage/Set-TableEntity.ps1]:
```suggestion
```

File: tasks/EnvironmentConfiguration/task/ps_modules/Handlers/private/storage/Set-TableEntity.ps1 [hunk 13]
```diff
@@ -0,0 +1,29 @@
...
+    }
+
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/task/ps_modules/Handlers/private/storage/Set-TableEntity.ps1]:
```suggestion
```

File: Build-Task.ps1 [hunk 14]
```diff
@@ -0,0 +1,243 @@
...
+    Packge and publish an extension package to the marketplace
+
+    .PARAMETER AccessToken
+    A personal access token with enough privileges to publish the package
+
+    .PARAMETER SkipRestore
+    Don't restore dependencies listed in dependency.json
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    ./Build-Task.ps1 -TaskRoot ./MyTask
```

File: tasks/EnvironmentConfiguration/tests/New-ConfigurationTableEntry.unit.test.ps1 [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+        $NewConfigurationTableEntryParameters = @{
+            SourcePath = "$PSScriptRoot/resource"
+            TargetFileName = "*.schema.json"
+            StorageAccount = "mock-storage-account"
+            EnvironmentName = "DEV"
+        }
+
+        Context "When passed a valid schema definition" {
+
+            It "Should sucesfully create a new configuration entity" {
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/tests/New-ConfigurationTableEntry.unit.test.ps1]:
```suggestion
            It "Should successfully create a new configuration entity" {
```

File: Build-Task.ps1 [hunk 16]
```diff
@@ -0,0 +1,243 @@
...
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask -Clean
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    ./Build-Task.ps1 -TaskRoot ./MyTask -Clean
```

File: Build-Task.ps1 [hunk 17]
```diff
@@ -0,0 +1,243 @@
...
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask -Clean -Build
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    ./Build-Task.ps1 -TaskRoot ./MyTask -Clean -Build
```

File: tasks/EnvironmentConfiguration/tests/Test-Configuration.unit.tests.ps1 [hunk 18]
```diff
@@ -0,0 +1,32 @@
...
+            Clear-MockEnvironment
+        }
+
+        $SchemaDefinition = "$PSScriptRoot/resource/SFA.DAS.Test.schema.json"
+        $ValidConfiguration = Get-Content -Path "$PSScriptRoot/resource/SFA.DAS.Test.Valid.json" -Raw
+        $InvalidConfiguration = Get-Content -Path "$PSScriptRoot/resource/SFA.DAS.Test.Invalid.json" -Raw
+
+        Context "When the configuration passed matches the schema definition" {
+
+            It "Should validate succesfully and not throw an exception" {
```
Comment (dalemorson) [inline tasks/EnvironmentConfiguration/tests/Test-Configuration.unit.tests.ps1]:
```suggestion
            It "Should validate successfully and not throw an exception" {
```

File: Build-Task.ps1 [hunk 19]
```diff
@@ -0,0 +1,243 @@
...
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask -Clean -Publish
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    ./Build-Task.ps1 -TaskRoot ./MyTask -Clean -Publish
```

File: Build-Task.ps1 [hunk 20]
```diff
@@ -0,0 +1,243 @@
...
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask -Publish -SkipRestore
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
    ./Build-Task.ps1 -TaskRoot ./MyTask -Publish -SkipRestore
```

File: Build-Task.ps1 [hunk 21]
```diff
@@ -0,0 +1,243 @@
...
+    }
+
+    $null = New-Item -Path $ReleaseTaskRoot -ItemType Directory -Force -ErrorAction SilentlyContinue
+    Copy-Item -Path $ResolvedTaskRoot -Destination "$PSScriptRoot/Release" -Recurse -Force
+
+    if (!$SkipRestore.IsPresent) {
+
+        foreach ($Package in $Config.Include | Sort-Object -Property Type) {
+            Write-Verbose -Message "Processing package dependency $($Package.Name)"
+            Write-Verbose -Message "Clean package directories: $($NoResotre.IsPresent)"
```
Comment (jack-education) [inline Build-Task.ps1]:
```suggestion
            Write-Verbose -Message "Clean package directories: $($Clean.IsPresent)"
```

File: Build-Task.ps1 [hunk 22]
```diff
@@ -0,0 +1,243 @@
...
+    Restore-TaskDependency -TaskRoot ./MyTask -Clean -Publish
+
+    .EXAMPLE
+    Restore-TaskDependency -TaskRoot ./MyTask -Publish -SkipRestore
+
+    .NOTES
+    Requirements:
+    The following applications must be installed an available in your PATH
+    - git
+    - tfx-cli
```
Comment (EwanNoble) [inline Build-Task.ps1]: Is tfx-cli still required for this script?

File: Build-Task.ps1 [hunk 23]
```diff
@@ -0,0 +1,243 @@
...
+            Write-Verbose -Message "Processing package dependency $($Package.Name)"
+            Write-Verbose -Message "Clean package directories: $($NoResotre.IsPresent)"
+
+            Write-Verbose -Message "Resolving package path"
+            [System.IO.FileInfo]$ResolvedPackagePath = "$($ReleaseTaskRoot)/$($Package.Path)"
+            Write-Verbose -Message "ResolvedPackagePath: $ReleaseTaskRoot"
+
+            switch ($Package.Type) {
+                'PSGallery' {
+
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
```

File: Build-Task.ps1 [hunk 24]
```diff
@@ -0,0 +1,243 @@
...
+                        $Package.Copy | ForEach-Object {
+                            Write-Host "[PSGallery] Copying module $($Package.Name) to $($Package.Path)"
+                            Copy-Item -Path $PackageInstallDirectory/$_ -Destination "$ResolvedPackagePath/$($Package.Name)" -Recurse -Force
+                        }
+                    }
+
+                    break
+                }
+                'Nuget' {
+
```
Comment (EwanNoble) [inline Build-Task.ps1]:
```suggestion
```

File: tasks/EnvironmentConfiguration/task/ps_modules/Handlers/Handlers.psm1 [hunk 25]
```diff
@@ -0,0 +1,15 @@
+# --- Dot source handlers
+Get-ChildItem -Path $PSScriptRoot -Filter *.ps1 -File  -Recurse | ForEach-Object {
+    . $_.FullName
+}
+
+Export-ModuleMember -Function @(
+    'New-ConfigurationTableEntry'
+)
+
+
```
Comment (EwanNoble) [inline tasks/EnvironmentConfiguration/task/ps_modules/Handlers/Handlers.psm1]:
```suggestion
```

File: tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1 [hunk 26]
```diff
@@ -0,0 +1,125 @@
...
+
+        Mock Get-AzureRmStorageAccountKey {
+            $MockKeysArray = @(
+                @{
+                    KeyName     = "key1"
+                    Value       = "bW9jayBzdG9yYWdlIGFjY291bnQga2V5IG5vdGhpbmcgdG8gc2VlIGhlcmUgMQ=="
+                    Permissions = "Full"
+                },
+                @{
+                    Key1        = "key2"
```
Comment (jack-education) [inline tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1]:
```suggestion
                    KeyName        = "key2"
```

File: tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1 [hunk 27]
```diff
@@ -0,0 +1,125 @@
...
+            It "Should throw if there is a failure when building a storage context" {
+                Mock New-AzureStorageContext {
+                    throw "Could not build context"
+                }
+
+                { New-ConfigurationEntity @NewConfigurationEntityParameters } | Should Throw
+            }
+        }
+
+        Context "When all parameters are correct and and there is an existing entity in the configuration table" {
```
Comment (jack-education) [inline tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1]:
```suggestion
        Context "When all parameters are correct and there is an existing entity in the configuration table" {
```

File: tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1 [hunk 28]
```diff
@@ -0,0 +1,125 @@
...
+                Assert-MockCalled -CommandName Get-AzureRmResource -Times 1
+                Assert-MockCalled -CommandName Get-AzureRmStorageAccountKey -Times 1
+                Assert-MockCalled -CommandName Get-AzureStorageTable -Times 1
+
+                Assert-MockCalled -CommandName Get-TableEntity -Times 1
+                Assert-MockCalled -CommandName Set-TableEntity -Times 1
+            }
+        }
+
+        Context "When all parameters are correct and and there is not an existing entity in the configuration table" {
```
Comment (jack-education) [inline tasks/EnvironmentConfiguration/tests/New-ConfigurationEntity.unit.tests.ps1]:
```suggestion
        Context "When all parameters are correct and there is not an existing entity in the configuration table" {
```

File: Build-Task.ps1 (hunk 3 above)
Comment (chelnak) [reply Build-Task.ps1]: i'll look to remove this

**Comment by EwanNoble** [comment]
@dalemorson @jack-education awaiting approval

**Pull request merged**